### PR TITLE
ci: check and retry Go package download

### PR DESCRIPTION
**Pull request by dzhu**
## Description

We had one CI failure [1] caused by the Go package download somehow fetching some incorrect file (seems like probably a 5xx status page or something, though there's no way to know now). This attempts to work around that by checking the hash of the downloaded file to at least cover for totally transient errors.

Also upgrade from 1.18 to 1.18.10 while we're here.

[1] https://app.circleci.com/pipelines/github/determined-ai/determined/33339/workflows/dd7d6e62-e6c8-4152-bb41-3d313ae9adab/jobs/1170863

## Test Plan

- [x] run script locally with correct and incorrect hashes
- [x] CI

### CODE REVIEW

File: .circleci/config.yml
Comment (dannysauer) [inline .circleci/config.yml:212]: You had a rare opportunity to use a do/until loop here, and missed it.  I'm so very disappointed. ;)

counter=3; until sha; do if $(( counter-- )); then echo fail; exit 1; curl; done

**Review by dannysauer — APPROVED** [review]
I'm good with this.

However, before mashing that merge button, do you know if the download failures are returning a code which would trigger curl's failure detection?  There's `curl --retry` and a host of associated retry options.  If so, staying with a single curl command would probably be preferable, despite my love of all things shell scripting.

Here's the curl doc on the subject: https://everything.curl.dev/usingcurl/downloads/retry

**Comment by dzhu** [comment]
Unfortunately, all I know is that the downloaded file was much too small (1613 bytes instead of ~100 MB) and, unsurprisingly, not a valid gzip file; it seems _likely_ that it was an error page, and _likely_ that it had an appropriate status code, but it is unknowable now.

That said, now that you've mentioned it... considering how rare this failure is, I think it would make sense to stick to the simpler `--retry`, even without being certain if that will work, and come back to this if it turns out that not to.

**Pull request merged**
